### PR TITLE
Add psel (playerSelect) support

### PR DESCRIPTION
**Pull request by XorTroll**
Everything was tested working fine (all modes and flags)

### CODE REVIEW

**Comment by yellows8** [comment]
Can you update this so that it matches sdknso better? See also pctlauth.

Guess this doesn't handle newer versions either, 9.0 has a newer commonargs-version. There's also an option for pushing an additional storage, so libappletLaunch can't be used for this.

**Comment by XorTroll** [comment]
What exactly do you mean with "match sdknso better"?
To have similar enum names, functions, descriptions...?
I'm looking on version values used (sdk uses 0x10000 but not sure for which version)

**Comment by XorTroll** [comment]
A newer sdk nso uses 0x20000 as version, so I'm assuming 9.0+ is 0x20000.

**Comment by yellows8** [comment]
Names + funcs. See friendsLa, sdknso handling seems to be similar-ish.

**Comment by XorTroll** [comment]
Still need to figure out for what is the larger storage used for (all the added UI modes don't seem to use it...)

**Comment by yellows8** [comment]
Yeah that LargeStorage doesn't seem to be used in sdksno (dunno if qlaunch etc uses that func?).

Also, if any of the impl'd functionality needs a certain sysver, there should be incompat-sysver checks for it.

File: nx/include/switch/applets/psel.h [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+ * @brief Wrapper for using playerSelect (user selection applet).
+ * @author XorTroll
+ * @copyright libnx Authors
+ */
+#pragma once
+#include "../types.h"
+#include "../services/applet.h"
+#include "../services/acc.h"
+
+// playerSelect UI modes.
```
Comment (yellows8) [inline nx/include/switch/applets/psel.h]: ///

File: nx/include/switch/applets/psel.h [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    PselUiMode_LicenseRequirementsForNetworkService = 10, ///< LicenseRequirementsForNetworkService
+    PselUiMode_NaLoginTest                          = 12, ///< NaLoginTest
+} PselUiMode;
+
+// UI settings for playerSelect.
+typedef struct {
+    u32 mode;                                     ///< UI mode, see \ref PselUiMode.
+    u32 dialogType;                               ///< Dialog type
+    u128 invalidUserList[ACC_USER_LIST_SIZE];     ///< List of user IDs which will be disabled.
+    u8 unk1[8];                                   ///< Unknown.
```
Comment (yellows8) [inline nx/include/switch/applets/psel.h]: `u8 unk_x{offset}[{hex}];`

File: nx/include/switch/applets/psel.h [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    u8 networkServiceRequired;                    ///< Whether the user needs to be linked to a Nintendo account.
+    u8 unk2[2];                                   ///< Unknown.
+    u8 allowUserCreation;                         ///< (With ::PselUiMode_SelectUser) enables the option to create a new user.
+    u8 skipEnabled;                               ///< Enables the option to skip user selection (a new button is shown)
+    u8 unk3[11];                                  ///< Unknown.
+} PselUiSettings;
+
+// Result data sent after execution.
+typedef struct {
+    u32 result;                      ///< Result code.
```
Comment (yellows8) [inline nx/include/switch/applets/psel.h]: Can use Result for this.

File: nx/include/switch/applets/psel.h [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    u128 userId;                     ///< Selected user ID.
```
Comment (yellows8) [inline nx/include/switch/applets/psel.h]: u128 is aligned to 0x10-bytes, not 8-bytes. userIDs (including this one) need the latter, see the handling elsewhere for that.

File: nx/source/applets/psel.c [hunk 5]
```diff
@@ -0,0 +1,66 @@
+#include <string.h>
+#include "types.h"
+#include "result.h"
+#include "services/applet.h"
+#include "services/set.h"
```
Comment (yellows8) [inline nx/source/applets/psel.c]: Guess this was supposed to be acc.h?

File: nx/source/applets/psel.c [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    LibAppletArgs args;
+    u32 la_ver = _pselGetLaVersion();
+    libappletArgsCreate(&args, la_ver);
+    PselResult res;
+    size_t reply_size;
+
+    rc = libappletLaunch(AppletId_playerSelect, &args, ui, sizeof(PselUiSettings), &res, sizeof(res), &reply_size);
+    
+    if (R_SUCCEEDED(rc)) {
+        if (R_FAILED(res.result)) rc = MAKERESULT(Module_Libnx, LibnxError_LibAppletBadExit);
```
Comment (yellows8) [inline nx/source/applets/psel.c]: `rc = res.result;`

File: nx/include/switch/applets/psel.h [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    PselUiMode_NicknameEditor                       = 4,  ///< NicknameEditor
+    PselUiMode_ForStarter                           = 5,  ///< Mode "starter" uses to register the console's first user on the initial setup
+    PselUiMode_NetworkServiceAccountRegistration    = 8,  ///< NetworkServiceAccountRegistration
+    PselUiMode_NintendoAccountNnidLinker            = 9,  ///< NintendoAccountNnidLinker
+    PselUiMode_LicenseRequirementsForNetworkService = 10, ///< LicenseRequirementsForNetworkService
+    PselUiMode_NaLoginTest                          = 12, ///< NaLoginTest
+} PselUiMode;
+
+// UI settings for playerSelect.
+typedef struct {
```
Comment (yellows8) [inline nx/include/switch/applets/psel.h]: Official sw stores structs within UiSettings, update this struct for that (one of those funcs also writes a titleID to UiSettings, FWIW).

File: nx/source/applets/psel.c (hunk 6 above)
Comment (XorTroll) [reply nx/source/applets/psel.c]: The result returned on the out data isn't actually a normal result I think, according to switchbrew
https://switchbrew.org/wiki/Profile_Selector

File: nx/include/switch/applets/psel.h (hunk 3 above)
Comment (XorTroll) [reply nx/include/switch/applets/psel.h]: As commented below, according to switchbrew the result returned on the out data isn't actually a normal result

File: nx/source/applets/psel.c (hunk 6 above)
Comment (yellows8) [reply nx/source/applets/psel.c]: Then don't use R_FAILED?

Comment (XorTroll) [reply nx/source/applets/psel.c]: Well on this sense it is also 0 on success and other value on failure it so the macro was suitable, if this leads to confusion won't use it I guess

File: nx/include/switch/applets/psel.h [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    PselUiMode_NintendoAccountNnidLinker            = 9,  ///< NintendoAccountNnidLinker
+    PselUiMode_LicenseRequirementsForNetworkService = 10, ///< LicenseRequirementsForNetworkService
+    PselUiMode_NaLoginTest                          = 12, ///< NaLoginTest
+} PselUiMode;
+
+/// UI settings for playerSelect.
+typedef struct {
+    u32 mode;                                                   ///< UI mode, see \ref PselUiMode.
+    u32 dialogType;                                             ///< Dialog type
+    union { u128 invalidUserList[ACC_USER_LIST_SIZE]; } PACKED; ///< List of user IDs which will be disabled.
```
Comment (fincs) [inline nx/include/switch/applets/psel.h]: Use AccountUid here too.

File: nx/include/switch/applets/psel.h [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+    u8 unk_x1[0x2];                                             ///< Unknown.
+    u8 allowUserCreation;                                       ///< (With ::PselUiMode_SelectUser) enables the option to create a new user.
+    u8 skipEnabled;                                             ///< Enables the option to skip user selection (a new button is shown)
+    u8 unk_x2[0xB];                                             ///< Unknown.
+} PselUiSettings;
+
+/// Result data sent after execution.
+typedef struct {
+    u32 result;                      ///< Result code.
+    union { u128 userId; } PACKED;   ///< Selected user ID.
```
Comment (fincs) [inline nx/include/switch/applets/psel.h]: Use AccountUid here too.

File: nx/include/switch/applets/psel.h [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+ */
+void pselUiSetSkipEnabled(PselUiSettings *ui, bool flag);
+
+/**
+ * @brief Shows the applet with the specified UI settings.
+ * @param ui PselUiSettings struct.
+ * @param out_uid Selected user ID.
+ * @note If user skips (see \ref pselUiSetSkipEnabled) this will return successfully but the output ID will be 0.
+ */
+Result pselUiShow(PselUiSettings *ui, u128 *out_user);
```
Comment (fincs) [inline nx/include/switch/applets/psel.h]: Use AccountUid here too.

File: nx/include/switch/applets/psel.h (hunk 9 above)
Comment (XorTroll) [reply nx/include/switch/applets/psel.h]: Ah, didn't know it has a custom type now

**Comment by XorTroll** [comment]
On another topic, sdkver 5.3.0 uses 0x10000 LaVersion, while 7.3.2 uses 0x20000 (unsure which console versions they are paired to)

**Comment by yellows8** [comment]
See the Doozer build log.

**Comment by XorTroll** [comment]
I haven't updated libnx for a while in my fork, is it the reason of that fail?

**Comment by yellows8** [comment]
yes

**Comment by fincs** [comment]
PR needs rebasing against libnx master.

**Comment by XorTroll** [comment]
Sdknso 8.2.0 also uses 0x20000 version, btw

File: nx/include/switch/applets/psel.h (hunk 7 above)
Comment (XorTroll) [reply nx/include/switch/applets/psel.h]: What exactly do you mean with "stores structs within UiSettings", that its content is split into different structs?

Comment (yellows8) [reply nx/include/switch/applets/psel.h]: Yes, same as official sw.

File: nx/source/applets/psel.c [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    for(i = 0; i < ACC_USER_LIST_SIZE; i++) {
+
+        if(!accountUidIsValid(&ui->invalidUserList[i])) {
+            __builtin_memcpy(&ui->invalidUserList[i], user_id, sizeof(AccountUid));
+            break;
+        }
+    }
+}
+
+void pselUiSetAllowUserCreation(PselUiSettings *ui, bool flag) {
```
Comment (fincs) [inline nx/source/applets/psel.c]: Should be a NX_CONSTEXPR function in the header

File: nx/source/applets/psel.c [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+    if(ui->mode == PselUiMode_SelectUser) {
+        ui->allowUserCreation = flag;
+    }
+}
+
+void pselUiSetNetworkServiceRequired(PselUiSettings *ui, bool flag) {
```
Comment (fincs) [inline nx/source/applets/psel.c]: Should be a NX_CONSTEXPR function in the header

File: nx/source/applets/psel.c [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+    ui->networkServiceRequired = flag;
+}
+
+void pselUiSetSkipEnabled(PselUiSettings *ui, bool flag) {
```
Comment (fincs) [inline nx/source/applets/psel.c]: Should be a NX_CONSTEXPR function

**Comment by yellows8** [comment]
An API that's semi-similar to official sw should also be used, like the rest of the library-applets. For example: `{prefix}ShowUserIconEditor(AccountUid uid)`

**Comment by XorTroll** [comment]
Added a simplified API on the latest commits

**Comment by yellows8** [comment]
Might adjust this later (?).

**Pull request merged**